ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -0,0 +1,592 @@
+      close();
+      throw new IOException("Encountered exception while parsing json, connecting to Helix or initializing", e);
+    }
+    helixClusterManagerMetrics.initializeInstantiationMetric(true);
+    helixClusterManagerMetrics.initializeDatacenterMetrics();
+    helixClusterManagerMetrics.initializeDataNodeMetrics();
+    helixClusterManagerMetrics.initializeDiskMetrics();
+    helixClusterManagerMetrics.initializePartitionMetrics();
+    helixClusterManagerMetrics.initializeCapacityMetrics();
+  }
+
+  /**
+   * Populate the initial data from the admin connection. Create nodes, disks, partitions and replicas for the entire
+   * cluster.
+   * @throws Exception if creation of {@link AmbryDataNode}s or {@link AmbryDisk}s throw an Exception.
+   */
+  private void initialize() throws Exception {
+    for (DcZkInfo dcZkInfo : dcToDcZkInfo.values()) {
+      HelixAdmin admin = dcZkInfo.helixManager.getClusterManagmentTool();
+      for (String instanceName : admin.getInstancesInCluster(clusterName)) {
+        InstanceConfig instanceConfig = admin.getInstanceConfig(clusterName, instanceName);
+        AmbryDataNode datanode = new AmbryDataNode(dcZkInfo.dcName, clusterMapConfig, instanceConfig.getHostName(),
+            Integer.valueOf(instanceConfig.getPort()), getRackId(instanceConfig), getSslPortStr(instanceConfig));
+        initializeDisksAndReplicasOnNode(datanode, instanceConfig);
+        instanceNameToDataNodeIdMap.put(instanceName, datanode);
+        dcZkInfo.clusterChangeListener.allInstances.add(instanceName);
+      }
+    }
+    for (Set<AmbryDisk> disks : dataNodeIdToDiskIds.values()) {
+      for (AmbryDisk disk : disks) {
+        clusterWideRawCapacityBytes += disk.getRawCapacityInBytes();
+      }
+    }
+    for (Set<AmbryReplica> partitionReplicas : partitionIdToReplicaIds.values()) {
+      long replicaCapacity = partitionReplicas.iterator().next().getCapacityInBytes();
+      clusterWideAllocatedRawCapacityBytes += replicaCapacity * partitionReplicas.size();
+      clusterWiseAllocatedUsableCapacityBytes += replicaCapacity;
+    }
+  }
+
+  /**
+   * Initialize the disks and replicas on the given node. Create partitions if this is the first time a replica of
+   * that partition is being constructed.
+   * @param datanode the {@link AmbryDataNode} that is being initialized.
+   * @param instanceConfig the {@link InstanceConfig} associated with this datanode.
+   * @throws Exception if creation of {@link AmbryDisk} throws an Exception.
+   */
+  private void initializeDisksAndReplicasOnNode(AmbryDataNode datanode, InstanceConfig instanceConfig)
+      throws Exception {
+    dataNodeIdToReplicaIds.put(datanode, new HashSet<AmbryReplica>());
+    dataNodeIdToDiskIds.put(datanode, new HashSet<AmbryDisk>());
+    List<String> sealedReplicas = getSealedReplicas(instanceConfig);
+    Map<String, Map<String, String>> diskInfos = instanceConfig.getRecord().getMapFields();
```

Comment:
could you explain how this works? What does `getMapFields()` do? Are all the map fields only disk info? What if we add a map field in the future that isn't disk info?